feat(commands)!: Add support for injected parameters autocompletions (#675)
DisnakeDev/disnake: changelog/670.feature.rst

|commands| Add support for injected parameters autocompletion

> 	Document :class:`.ext.commands.Injection`

Author: Fixed in [9136fcf412f032a0268c589a541efdcd819844fb](https://github.com/DisnakeDev/disnake/pull/675/commits/9136fcf412f032a0268c589a541efdcd819844fb)

	:meth:`.ext.commands.Injection.register` now returns :class:`.ext.commands.Injection`

> 	:func:`.ext.commands.register_injection` now returns :class:`.ext.commands.Injection`

Reviewer: I'd argue these two return type changes should be marked as breaking, just in case. @onerandomusername thoughts?

Author: I guess thoss return values weren't meant for public use previously (i.e., not a part of a public api) and thus this change shouldn't be considered breaking
There also was a related conversation on discord: https://discord.com/channels/808030843078836254/913779868985090089/1002217259374227467

Reviewer 2: At least changing `register_injection` is breaking, but `Injection.register` was typed as returning `Any`. IMO if moving `register_injection` to a breaking changelog entry, we should move `Injection.register` as well.

Author: So, should we consider these changes breaking?

Reviewer: We should mark them as breaking, I'd rather have something labeled as breaking without it actually breaking anything, than the other way around.
The runtime impact could be avoided by making `Injection` generic and adding a `__call__` that passes calls through to `self.function`, such that this would basically end up only being a breaking change in terms of typing.

Author: Im too lazy to reference commits, but done

> |commands| Add support for injected parameters autocompletion
> - Document :class:`.ext.commands.Injection`
> - :meth:`Injection.register <ext.commands.Injection.register>` now returns :class:`Injection <ext.commands.Injection>`
> - :func:`.ext.commands.register_injection` now returns :class:`.ext.commands.Injection`
> - Add decorator interface for :func:`.ext.commands.inject`: :func:`.ext.commands.injection`
> - Add :meth:`.ext.commands.Injection.autocomplete` decorator
> - Add ``autocompleters`` keyword-only argument to :class:`.ext.commands.Injection` constructor
> - Add ``autocompleters`` keyword-only argument to :func:`.ext.commands.inject`
> - Add ``autocompleters`` keyword-only argument to :func:`.ext.commands.register_injection` constructor

Reviewer: ```suggestion
|commands| Add support for injected parameters autocompletion
- Add :meth:`Injection.autocomplete <ext.commands.Injection.autocomplete>` decorator
- :func:`register_injection <ext.commands.register_injection>` now returns an :class:`Injection <ext.commands.Injection>` object
- Add :func:`injection <ext.commands.injection>` as a decorator interface for :func:`inject <ext.commands.inject>`
- Add ``autocompleters`` keyword-only argument to :class:`Injection <ext.commands.Injection>`, :func:`inject <ext.commands.inject>`, and :func:`register_injection <ext.commands.register_injection>`
```
couple minor changes:
- removed `ext.commands` prefix - the raw rst is a bit more cumbersome to write, but the end result is easier to read
- removed line about `Injection.register`, not documented as per previous conversation
- merged last three lines into one

The part about `Injection` now being documented should ideally be in a separate `670.doc.rst` changelog entry.

- Add :meth:`Injection.autocomplete <ext.commands.Injection.autocomplete>` decorator
- Add :func:`injection <ext.commands.injection>` as a decorator interface for :func:`inject <ext.commands.inject>`
- Add ``autocompleters`` keyword-only argument to :class:`Injection <ext.commands.Injection>`, :func:`inject <ext.commands.inject>`, and :func:`register_injection <ext.commands.register_injection>`
- Add :attr:`SubCommand.parent <ext.commands.SubCommand.parent>`
- Add :attr:`SubCommandGroup.parent <ext.commands.SubCommandGroup.parent>`

> - Add :attr:`SubCommand.parent_command <ext.commands.SubCommand.parent_command>` property

Reviewer 2: these should be in a separate file. Ideally we should move these changes to a new pr and merge that first, rather than doing both at once.

Author: @EQUENOS Would you want to do this yourself or should i do this?

Reviewer 3: I can do this

Reviewer 2: resolved as of #759
